plutus-core/plutus-core/src/PlutusCore/Default/Universe.hs

```diff
@@ -131,68 +134,146 @@ pattern DefaultUniArray uniA =
 pattern DefaultUniPair uniA uniB =
     DefaultUniProtoPair `DefaultUniApply` uniA `DefaultUniApply` uniB
 
+-- Removing 'LoopBreaker' didn't change anything at the time this comment was written, but we kept
+-- it, because it hopefully provides some additional assurance that 'geqL' will not get elaborated
+-- as a recursive definition.
+instance AllBuiltinArgs DefaultUni (GEqL DefaultUni) a => GEqL DefaultUni a where
+    geqL DefaultUniInteger a2 = do
+        DefaultUniInteger <- pure a2
+        pure Refl
+    geqL DefaultUniByteString a2 = do
+        DefaultUniByteString <- pure a2
+        pure Refl
+    geqL DefaultUniString a2 = do
+        DefaultUniString <- pure a2
+        pure Refl
+    geqL DefaultUniUnit a2 = do
+        DefaultUniUnit <- pure a2
+        pure Refl
+    geqL DefaultUniBool a2 = do
+        DefaultUniBool <- pure a2
+        pure Refl
+    geqL (DefaultUniProtoList `DefaultUniApply` a1) listA2 = do
+        DefaultUniProtoList `DefaultUniApply` a2 <- pure listA2
+        Refl <- geqL (LoopBreaker a1) (LoopBreaker a2)
```

Comment:
Can you not apply the same `LoopBreaker` trick to `geq`?

Reply:
It's not `LoopBreaker` that is the trick. It's turning term-level recursion into type-level recursion. `LoopBreaker` is just to make it harder for GHC to see through the actual trick. You can see a comment above this instance, which says that `LoopBreaker` doesn't actually seem to impact anything, I just put it in there to increase the odds GHC won't rebel.

Reply:
I probably should re-review this then. I thought by having `LoopBreaker`, the `geqL` here is no longer a recursive call, but it calls the `geqL` of a different instance - `LoopBreaker`'s, hence it makes `geqL` inlinable. But that's not how it works?